File: python/003.py
'''
https://projecteuler.net/problem=3

Largest prime factor

The prime factors of 13195 are 5, 7, 13 and 29.
What is the largest prime factor of the number 600851475143 ?

Status: Complete
'''

# Returns true if x is a prime number
def isPrime(x):
    # throw out any even numbers
    if x % 2 == 0:
        return False
    i = int(round(x/3)+1) # Highest prime factor of a number cannot be higher than 1/3 of that number, so start with x / 3.
    while i > 2:
        if x % i == 0:
            return False
        i = i - 1
    return True

# Returns highest prime factor of 600851475143 under a limit
def calc(limit):
    i = limit
    while limit > 0:
        if 600851475143 % i == 0:
            if isPrime(i):
                return i
        i = i - 1
    return -1

print(calc(100000))

File: java/003.java
/**
 * https://projecteuler.net/problem=3
 * 
 * The prime factors of 13195 are 5, 7, 13 and 29.
 * What is the largest prime factor of the number 600851475143 ?
 * 
 * Status: Complete
 */

import java.util.*;

public class eulerP3 {    
    public static List<Integer> primes = new ArrayList<Integer>();
    public static boolean isPrime=true;
    public static long num = 600851475143L;
    public static int limit = 100000;
    public static int lpf = 0;

    public static void populatePrimeList(int j){
        // populates an arraylist of all prime numbers up to j
        primes.add(2);
        for (int i = 3; i <= j; i+=2){
            for (int k=2; k<i; k++) {
                if (i%k==0) {
                    isPrime=false;
                }
            }
            if (isPrime){
                primes.add(i);
            }
            isPrime = true;
        }
    }

    public static void main(String [] args) {
        populatePrimeList(limit);

        //         for (int i : primes){
        //             System.out.println(i);
        //         }
        
        for (int i = primes.size()-1; i > 0; i--){
            int j = primes.get(i);
            if (num%j==0){
                lpf = j;
                break;
            }
        }
        System.out.print("Largest prime factor = "+lpf);
    }
}
